### PR TITLE
Groups not in the structure should not visible be in the user details

### PR DESCRIPTION
**Pull request by simbas**
Related to issue CAV2-371: https://opendigitaleducation.atlassian.net/browse/CAV2-371

### CODE REVIEW

File: admin/src/main/ts/app/users/details/sections/manualgroups/user-manualgroups-section.component.ts [hunk 1]
```diff
@@ -47,63 +52,69 @@ import { AbstractSection } from '../abstract.section'
...
+                .filter(group => isGroupManageable(group, this.globalStore));
+        } else {
+            this.filteredGroups = [];
+        }
     }
 
     private updateLightboxManualGroups() {
         this.lightboxManualGroups = this.structure.groups.data.filter(
-            g => g.type === 'ManualGroup' 
+            g => g.type === 'ManualGroup'
```
Comment (Chokoneko) [inline admin/src/main/ts/app/users/details/sections/manualgroups/user-manualgroups-section.component.ts]: nommage de g et mg

File: admin/src/main/ts/app/users/details/sections/manualgroups/user-manualgroups-section.component.ts [hunk 2]
```diff
@@ -47,63 +52,69 @@ import { AbstractSection } from '../abstract.section'
...
                 && !this.details.manualGroups.find(mg => mg.id == g.id));
     }
 
-    filterByInput = (mg: {id: string, name: string}) => {
+    filterByInput(mg: { id: string, name: string }): boolean {
```
Comment (Chokoneko) [inline admin/src/main/ts/app/users/details/sections/manualgroups/user-manualgroups-section.component.ts]: nommage de mg

File: admin/src/main/ts/app/core/store/models/user.model.ts [hunk 3]
```diff
@@ -1,158 +1,182 @@
...
+    classes: Classe[];
+    duplicates: { id: string, firstName: string, lastName: string, code: string, structures: string[] }[];
+    deleteDate?: number;
+    disappearanceDate?: number;
+
+    userDetails: UserDetailsModel;
 
     visibleStructures() {
-        return this.structures.filter(s => globalStore.structures.data.find(struct => struct.id === s.id))
+        return this.structures.filter(s => globalStore.structures.data
```
[the remark this answers is not in the document]
Comment (simbas) [reply admin/src/main/ts/app/core/store/models/user.model.ts]: done

File: admin/src/main/ts/app/core/store/models/user.model.ts [hunk 4]
```diff
@@ -1,158 +1,182 @@
...
+            .find(struct => struct.id === s.id));
     }
 
     invisibleStructures() {
-        return this.structures.filter(s => globalStore.structures.data.every(struct => struct.id !== s.id))
+        return this.structures.filter(s => globalStore.structures.data
```
[the remark this answers is not in the document]
Comment (simbas) [reply admin/src/main/ts/app/core/store/models/user.model.ts]: done

File: admin/src/main/ts/app/users/details/sections/manualgroups/user-manualgroups-section.component.ts [hunk 5]
```diff
@@ -47,63 +52,69 @@ import { AbstractSection } from '../abstract.section'
...
+    public user: UserModel;
+    @Input()
+    public structure: StructureModel;
 
-    @Input() user: UserModel;
-    @Input() structure: StructureModel;
-    
     constructor(
         public spinner: SpinnerService,
         private ns: NotifyService,
```
[the remark this answers is not in the document]
Comment (simbas) [reply admin/src/main/ts/app/users/details/sections/manualgroups/user-manualgroups-section.component.ts]: done

File: admin/src/main/ts/app/users/details/sections/manualgroups/user-manualgroups-section.component.ts [hunk 6]
```diff
@@ -47,63 +52,69 @@ import { AbstractSection } from '../abstract.section'
...
-        super()
-    }
-    
-    private _inputFilter = "";
-    set inputFilter(filter: string) {
-        this._inputFilter = filter;
-    }
-    get inputFilter() {
-        return this._inputFilter;
+        private cdRef: ChangeDetectorRef,
```
[the remark this answers is not in the document]
Comment (simbas) [reply admin/src/main/ts/app/users/details/sections/manualgroups/user-manualgroups-section.component.ts]: done

File: admin/src/main/ts/app/users/details/sections/manualgroups/user-manualgroups-section.component.ts (hunk 1 above)
Comment (simbas) [reply admin/src/main/ts/app/users/details/sections/manualgroups/user-manualgroups-section.component.ts]: done

File: admin/src/main/ts/app/users/details/sections/manualgroups/user-manualgroups-section.component.ts (hunk 2 above)
Comment (simbas) [reply admin/src/main/ts/app/users/details/sections/manualgroups/user-manualgroups-section.component.ts]: done

File: admin/src/main/ts/app/core/store/models/user.model.ts [hunk 7]
```diff
@@ -1,158 +1,182 @@
...
+        } catch (e) {
+            return {id: duplicate.id, structure: duplicate.structures[0]};
         }
     }
 
     separateDuplicate(duplicateId: string) {
         return this.http.delete(`/directory/duplicate/ignore/${this.id}/${duplicateId}`).then(() => {
-            let duplicate = this.duplicates.find(d => d.id === duplicateId)
+            let duplicate = this.duplicates.find(d => d.id === duplicateId);
             duplicate.structures.forEach(s => {
```
[the remark this answers is not in the document]
Comment (simbas) [reply admin/src/main/ts/app/core/store/models/user.model.ts]: done

File: admin/src/main/ts/app/core/store/models/user.model.ts [hunk 8]
```diff
@@ -1,158 +1,182 @@
...
-                let structure = globalStore.structures.data.find(struct => struct.id === s)
-                if(structure && structure.users.data.length > 0) {
-                    let dup = structure.users.data.find(u => u.id === duplicateId)
-                    if(dup) dup.duplicates = dup.duplicates.filter(d => d.id !== this.id)
+                let structure = globalStore.structures.data.find(struct => struct.id === s);
+                if (structure && structure.users.data.length > 0) {
+                    let dup = structure.users.data.find(u => u.id === duplicateId);
```
[the remark this answers is not in the document]
Comment (simbas) [reply admin/src/main/ts/app/core/store/models/user.model.ts]: done

File: admin/src/main/ts/app/users/details/sections/classes/user-classes-section.component.ts [hunk 9]
```diff
@@ -31,17 +36,18 @@ import { OnChanges, OnInit } from '@angular/core/src/metadata/lifecycle_hooks';
                     </list>
                 </div>
             </lightbox>
-            
+
             <ul class="actions-list">
-                <li *ngFor="let c of user?.classes">
+                <li *ngFor="let c of filteredClasses">
```
[the remark this answers is not in the document]
Comment (simbas) [reply admin/src/main/ts/app/users/details/sections/classes/user-classes-section.component.ts]: done

File: admin/src/main/ts/app/users/details/sections/classes/user-classes-section.component.ts [hunk 10]
```diff
@@ -97,22 +105,24 @@ export class UserClassesSection extends AbstractSection implements OnInit, OnCha
...
+    filterByInput(c: { id: string, name: string }): boolean {
+        if (!this.inputFilter) {
+            return true;
+        }
+        return `${c.name}`.toLowerCase().indexOf(this.inputFilter.toLowerCase()) >= 0;
     }
 
-    filterClasses = (c: {id: string, name: string}) => {
-        return !this.user.classes.find(classe => c.id === classe.id)
+    filterClasses(c: { id: string, name: string }): boolean {
```
[the remark this answers is not in the document]
Comment (simbas) [reply admin/src/main/ts/app/users/details/sections/classes/user-classes-section.component.ts]: done

File: admin/src/main/ts/app/users/details/sections/classes/user-classes-section.component.ts [hunk 11]
```diff
@@ -144,19 +154,19 @@ export class UserClassesSection extends AbstractSection implements OnInit, OnCha
             }, 'notify.user.remove.head.teacher.error.title', err)
         })
     }
-    
-    disableClass = (c) => {
+
+    disableClass(c) {
```
[the remark this answers is not in the document]
Comment (simbas) [reply admin/src/main/ts/app/users/details/sections/classes/user-classes-section.component.ts]: done

File: admin/src/main/ts/app/users/details/sections/functionalgroups/user-functionalgroups-section.component.ts [hunk 12]
```diff
@@ -30,14 +35,14 @@ import { SpinnerService, NotifyService } from '../../../../core/services'
                     </list>
                 </div>
             </lightbox>
-            
+
             <ul class="actions-list">
-                <li *ngFor="let g of details?.functionalGroups">
+                <li *ngFor="let g of filteredGroups">
```
[the remark this answers is not in the document]
Comment (simbas) [reply admin/src/main/ts/app/users/details/sections/functionalgroups/user-functionalgroups-section.component.ts]: done

File: admin/src/main/ts/app/users/details/sections/functionalgroups/user-functionalgroups-section.component.ts [hunk 13]
```diff
@@ -48,61 +53,70 @@ import { SpinnerService, NotifyService } from '../../../../core/services'
...
     }
 
     // Refresh data when structure change
     ngOnChanges() {
         this.updateLightboxFunctionalGroups();
+        this.filterManageableGroups();
+    }
+
+    private filterManageableGroups() {
+        if (this.details) {
```
[the remark this answers is not in the document]
Comment (simbas) [reply admin/src/main/ts/app/users/details/sections/functionalgroups/user-functionalgroups-section.component.ts]: c'est au jugé par rapport à la lisibilité

File: admin/src/main/ts/app/users/details/sections/functionalgroups/user-functionalgroups-section.component.ts [hunk 14]
```diff
@@ -48,61 +53,70 @@ import { SpinnerService, NotifyService } from '../../../../core/services'
...
+            this.filteredGroups = [];
+        }
     }
-    
+
     private updateLightboxFunctionalGroups() {
-        this.lightboxFunctionalGroups = this.structure.groups.data.filter(
-            g => g.type === 'FunctionalGroup' 
+        this.lightboxFunctionalGroups = this.structure.groups.data
+            .filter(g => g.type === 'FunctionalGroup'
```
[the remark this answers is not in the document]
Comment (simbas) [reply admin/src/main/ts/app/users/details/sections/functionalgroups/user-functionalgroups-section.component.ts]: done

File: admin/src/main/ts/app/users/details/sections/functionalgroups/user-functionalgroups-section.component.ts [hunk 15]
```diff
@@ -48,61 +53,70 @@ import { SpinnerService, NotifyService } from '../../../../core/services'
...
                 && !this.details.functionalGroups.find(mg => mg.id == g.id)
-        );
+            );
     }
 
-    filterByInput = (g: {id: string, name: string}) => {
-        if (!this.inputFilter) return true
-        return `${g.name}`.toLowerCase().indexOf(this.inputFilter.toLowerCase()) >= 0
+    filterByInput(g: { id: string, name: string }) {
```
[the remark this answers is not in the document]
Comment (simbas) [reply admin/src/main/ts/app/users/details/sections/functionalgroups/user-functionalgroups-section.component.ts]: done

**Pull request merged**